sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -1311,11 +1325,27 @@ class Dataset[T] private[sql](
       if (sqlContext.conf.supportQuotedRegexColumnName) {
         colRegex(colName)
       } else {
-        val expr = resolve(colName)
-        Column(expr)
+        Column(addDataFrameIdToCol(resolve(colName)))
       }
   }
 
+  // Attach the dataset id and column position to the column reference, so that we can detect
+  // ambiguous self-join correctly.See the rule `DetectAmbbiguousSelfJoin`.
+  // This must be called before we return a `Column` that contains `AttributeReference`.
```

Comment:
Also, document these metadata will be removed in the rule DetectAmbbiguousSelfJoin